### PR TITLE
User DAO Implementation 

### LINKED ISSUE #11: Extend data access object interfaces
Extending the data access object interfaces to interact with datastore.

### PR DESCRIPTION
**Pull request by nnenna352**
User dao implementation on UserDaoDatastore file and UserDao file 
Might write more tests and update accordingly

### CODE REVIEW

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        userEntity.setProptery("favoriteSDKs",user.favoriteSDKs);
+
+        DATASTORE.put(userEntity);
+    }
+ 
+    // Updates user favorite SDKs
+    public void updateUser(User user){
+         User outdated = getUser(user.uid());
+         outdated.favoriteSDKs() = user.favoriteSDKs();
+
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: you will have to put the updated user entity back into the datastore

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    }
+    
+
+
+
+
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: nit: delete the whitespace

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    public static final DatastoreService DATASTORE = DatastoreServiceFactory.getDatastoreService();
+
+    // Get a user from the datastore
+    public User getUser(String id){
+        Query query = new Query("User");
+        PreparedQuery results = DATASTORE.prepare(query);
+        QueryResultIterable<Entity> users = preparedQuery.asQueryResultIterable();
+
+        for (Entity user: users){
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: This works but it is probably better to let the datastore filter the user. Check out my getSDKs function on line 70 in this branch as well as the makePropertyFilter function on line 280 https://github.com/googleinterns/step104-2020/blob/SDKDaoImpl/src/main/java/firebase/sdk/dashboard/dao/SDKDaoDatastore.java

**Review by eriyezab — CHANGES_REQUESTED** [review]

File: src/main/java/firebase/sdk/dashboard/dao/UserDao.java [hunk 4]
```diff
@@ -7,7 +7,7 @@
 /**
  * An interface for retrieving and manipulating data about Users.
  */
-public interface UserDao {
+ public interface UserDao {
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDao.java]: Why the extra space at the start?

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.QueryResultIterable;
+
+/* Class to implement UserDao */
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: Please add more details like what kind of implementation this is.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java (hunk 3 above)
Comment (ankitaj224) [reply src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: +1

File: src/main/webapp/css/style.css [hunk 6]
```diff
@@ -1,30 +0,0 @@
-.favourite {
```
Comment (ankitaj224) [inline src/main/webapp/css/style.css]: Reason for deleting this file?
Seems unrelated change

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java (hunk 2 above)
Comment (ankitaj224) [reply src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: +1 and fix formatting

File: src/main/java/firebase/sdk/dashboard/dao/UserDao.java [hunk 7]
```diff
@@ -27,7 +27,10 @@
   /**
    * Registers the changes in the User object to the database.
    *
-   * @param user The user being update in the data source.
+   * @param user The user being updated in the data source.
    */
   public void updateUser(User user);
+  
 }
+
```
Comment (ankitaj224) [inline src/main/java/firebase/sdk/dashboard/dao/UserDao.java]: remove extra lines

**Review by ankitaj224 — CHANGES_REQUESTED** [review]

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    public static final DatastoreService DATASTORE = DatastoreServiceFactory.getDatastoreService();
+
+    // Get a user from the datastore
+    public User getUser(String id){
+         FilterPredicate userPropertyFilter = makePropertyFilter("user", id;
+        Query query = new Query("User").setFilter(userPropertyFilter);
+
+        PreparedQuery preparedQuery = DATASTORE.prepare(query);
+        FetchOptions fetchOptions = FetchOptions.Builder.withDefaults();
+        QueryResultIterable<Entity> users = preparedQuery.asQueryResultIterable();
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: nit: since you are only searching for a single user, you could use asSingleEntity() isntead of asQueryResultIterable(). That way you don't have to do a for loop

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        for (Entity entity : results) {
+            return entity;
+        }
+    }
+
+    // Add a user to the datastore
+    public void addUser(User user){
+       Key userKey = KeyFactory.createKey(user.uid());
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: The first argument should be "User" so that the datastore stores these entities as User entities

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+/* Class to implement UserDao */
+
+public class UserDaoDatastore implements UserDao {
+
+    public static final DatastoreService DATASTORE = DatastoreServiceFactory.getDatastoreService();
+
+    // Get a user from the datastore
+    public User getUser(String id){
+         FilterPredicate userPropertyFilter = makePropertyFilter("user", id;
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: instead of "user" it should be "id" since that is the property you are trying to filter on. It also seems like this line is missing a closing ")"

**Review by eriyezab — CHANGES_REQUESTED** [review]

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    // Get a user from the datastore
+    public User getUser(String id){
+         FilterPredicate userPropertyFilter = makePropertyFilter("id", id);
+        Query query = new Query("User").setFilter(userPropertyFilter);
+
+        PreparedQuery preparedQuery = DATASTORE.prepare(query);
+        FetchOptions fetchOptions = FetchOptions.Builder.withDefaults();
+        Entity user = preparedQuery.asSingleEntity();
+        if (user == null) {
+        return null;
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: nit: indentation

File: target/maven-archiver/pom.properties [hunk 12]
```diff
@@ -0,0 +1,5 @@
+#Generated by Maven
+#Tue Jul 21 17:17:50 UTC 2020
+groupId=firebase.sdk.dashboard
+artifactId=sdkdashboard
+version=1
```
Comment (eriyezab) [inline target/maven-archiver/pom.properties]: nit: this file and the whole target directory should be deleted

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        return user;
+    }
+
+    // Add a user to the datastore
+    public void addUser(User user){
+       Key userKey = KeyFactory.createKey(user);
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: change to 'Key userKey = KeyFactory.createKey("User", user.uid());'

this makes it so that the kind of the entity is "User" and its key value will be its uid

**Review by eriyezab — CHANGES_REQUESTED** [review]

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        Entity user = preparedQuery.asSingleEntity();
+        if (user == null) {
+            return null;
+        }
+        return user;
+    }
+
+    // Add a user to the datastore
+    public void addUser(User user){
+       Key userKey = KeyFactory.createKey("User",user.uid());
```
Comment (eriyezab) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: nit: add space after comma

**Review by eriyezab — APPROVED** [review]

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.QueryResultIterable;
+
+/* Class to implement UserDao */
+
```
Comment (ankitaj224) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: Remove this extra line. Use the google-java-formatter.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    // Get a user from the datastore
+    public User getUser(String id){
+        FilterPredicate userPropertyFilter = makePropertyFilter("id", id);
+        Query query = new Query("User").setFilter(userPropertyFilter);
+
+        PreparedQuery preparedQuery = DATASTORE.prepare(query);
+        FetchOptions fetchOptions = FetchOptions.Builder.withDefaults();
+        Entity user = preparedQuery.asSingleEntity();
+        if (user == null) {
+            return null;
```
Comment (ankitaj224) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java:41]: Add a TODO to throw user not found exception

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java (hunk 15 above)
Comment (ankitaj224) [reply src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: The whole file needs formatting.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        return user;
+    }
+
+    // Add a user to the datastore
+    public void addUser(User user){
+       Key userKey = KeyFactory.createKey("User",user.uid());
+       
+        // Create a user entity
+        Entity userEntity = new Entity(userKey);
+        userEntity.setProperty("id",user.uid());
```
Comment (ankitaj224) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: Did you check if its recommended to use user id or user token?

**Review by ankitaj224 — CHANGES_REQUESTED** [review]
Did you intend to delete styles.css?

File: src/main/java/firebase/sdk/dashboard/dao/UserDao.java [hunk 18]
```diff
@@ -27,7 +27,7 @@
   /**
    * Registers the changes in the User object to the database.
    *
-   * @param user The user being update in the data source.
+   * @param user The user being updated in the data source.
    */
-  public void updateUser(User user);
+  public void updateUser(User user);
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDao.java]: Remove spaces at the end of line.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+        userEntity.setProptery("favoriteSDKs",user.favoriteSDKs);
+
+        DATASTORE.put(userEntity);
+    }
+ 
+    // Updates user favorite SDKs
+    public void updateUser(User user){
+         User outdated = getUser(user.uid());
+         outdated.favoriteSDKs() = user.favoriteSDKs();
+         addUser(User outdated);
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: This will not compile. Remove User and also try compiling the code to catch more such issues.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+        userEntity.setProperty("email",user.email());
+        userEntity.setProptery("favoriteSDKs",user.favoriteSDKs);
+
+        DATASTORE.put(userEntity);
+    }
+ 
+    // Updates user favorite SDKs
+    public void updateUser(User user){
+         User outdated = getUser(user.uid());
+         outdated.favoriteSDKs() = user.favoriteSDKs();
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: Why copy all the parameters to another User object? I think this method should save whatever data is given in the input User object.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+    // Add a user to the datastore
+    public void addUser(User user){
+       Key userKey = KeyFactory.createKey("User",user.uid());
+       
+        // Create a user entity
+        Entity userEntity = new Entity(userKey);
+        userEntity.setProperty("id",user.uid());
+        userEntity.setProperty("email",user.email());
+        userEntity.setProptery("favoriteSDKs",user.favoriteSDKs);
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: -> favoriteSDKs()

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java (hunk 15 above)
Comment (shrutip90) [reply src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: https://github.com/google/google-java-format. You can install and run this to format easily.

**Review by shrutip90 — CHANGES_REQUESTED** [review]

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java (hunk 17 above)
Comment (nnenna352) [reply src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: We never store the user token, I don't think so at least

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.QueryResultIterable;
+
+/* Class to implement UserDao to manipulate data about the users */
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: nit: full stop at the end. Please add that to the other comments in this file as well.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+public class UserDaoDatastore implements UserDao {
+
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: nit: Remove blank line

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+        return users;
+    }
+
+    // Add a user to the datastore
+    public void addUser(User user){
+       Key userKey = KeyFactory.createKey("User", user.uid());
+       
+        // Create a user entity
+        Entity userEntity = new Entity(userKey);
+        userEntity.setProperty("id",user.uid());
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: Add space after comma.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+        userEntity.setProperty("email",user.email());
+        userEntity.setProptery("favoriteSDKs",user.favoriteSDKs());
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: ditto

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+
+        DATASTORE.put(userEntity);
+    }
+ 
+    // Updates user favorite SDKs
+    public void updateUser(User user){
+         FilterPredicate userPropertyFilter = makePropertyFilter("id", user.uid());
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: Remove space at the line start.

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+        Query query = new Query("User").setFilter(userPropertyFilter);
+
+        PreparedQuery preparedQuery = DATASTORE.prepare(query);
+        FetchOptions fetchOptions = FetchOptions.Builder.withDefaults();
+        Entity userEntity = preparedQuery.asSingleEntity();
+        if (userEntity == null) {
+            //TODO: Throw user not found exception
+        }
+         userEntity.setProperty("favoriteSDKs",user.favoriteSDKs());
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: ditto

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+         DATASTORE.put(userEntity);
+    }
+                                                                 
+   private FilterPredicate makePropertyFilter(String property, Object value) {
+        FilterPredicate propertyFilter = new FilterPredicate(property, FilterOperator.EQUAL, value);
+
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: Remove blank line

File: src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+        FetchOptions fetchOptions = FetchOptions.Builder.withDefaults();
+        Entity userEntity = preparedQuery.asSingleEntity();
+        if (userEntity == null) {
+            //TODO: Throw user not found exception
+        }
+         userEntity.setProperty("favoriteSDKs",user.favoriteSDKs());
+         DATASTORE.put(userEntity);
+    }
+                                                                 
+   private FilterPredicate makePropertyFilter(String property, Object value) {
```
Comment (shrutip90) [inline src/main/java/firebase/sdk/dashboard/dao/UserDaoDatastore.java]: You could name this equalPropertyFilter to indicate that the filter will add an equality condition.

**Review by shrutip90 — APPROVED** [review]

**Review by ankitaj224 — APPROVED** [review]
